### ISSUE TITLE
Private network IP assigned to wrong interface on vagrant reload

### ISSUE THREAD

**Issue by jamiejackson** [hashicorp/vagrant#9546]
# Short story

* vagrant up: everything looks good, private network ip assigned to eth1, as expected
* vagrant reload: can still reach the guest by private ip (everything still _behaves_ normally), but the nic configs show that eth0 also has the ip that eth1 has
* vagrant reload 2: eth0 still has the ip that eth1 is supposed to have, but the private interface (eth1) is down, so the private network is non-functional

# Info

* I don't know if this is relevant, but I'm using the docker provisioner (not the provider) and the stack has multiple overlay networks defined.
* Vagrant 2.0.2
* VirtualBox 5.2.8
* Guest Box: `ubuntu/trusty64 (virtualbox, 20180222.1.1)` with updated guest extensions
* Host: Mac 10.12.6
* Vagrantfile private network definition: `config.vm.network "private_network", ip: 192.168.56.10`

# Other info

* My team cannot reproduce this problem in Windows 10

# Repeatable experiment

## initial vagrant up

site up. nics look as expected, and so does `/etc/network/interfaces`

```
$ vagrant ssh -- ip a

2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 08:00:27:d1:30:c3 brd ff:ff:ff:ff:ff:ff
    inet 10.0.2.15/24 brd 10.0.2.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::a00:27ff:fed1:30c3/64 scope link
       valid_lft forever preferred_lft forever
3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 08:00:27:82:2b:6b brd ff:ff:ff:ff:ff:ff
    inet 192.168.56.10/24 brd 192.168.56.255 scope global eth1
       valid_lft forever preferred_lft forever
    inet6 fe80::a00:27ff:fe82:2b6b/64 scope link
       valid_lft forever preferred_lft forever
```

```
$ vagrant ssh -- cat /etc/network/interfaces

...
#VAGRANT-BEGIN
# The contents below are automatically generated by Vagrant. Do not modify.
auto eth1
iface eth1 inet static
      address 192.168.56.10
      netmask 255.255.255.0
#VAGRANT-END
```

## vagrant reload

site up. nics don't look as expected (now eth0 has the ip, too), and `/etc/network/interfaces` has the wrong nic specified.

```
$ vagrant ssh -- ip a

2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 08:00:27:d1:30:c3 brd ff:ff:ff:ff:ff:ff
    inet 10.0.2.15/24 brd 10.0.2.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet 192.168.56.10/24 brd 192.168.56.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::a00:27ff:fed1:30c3/64 scope link
       valid_lft forever preferred_lft forever
3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 08:00:27:82:2b:6b brd ff:ff:ff:ff:ff:ff
    inet 192.168.56.10/24 brd 192.168.56.255 scope global eth1
       valid_lft forever preferred_lft forever
    inet6 fe80::a00:27ff:fe82:2b6b/64 scope link
       valid_lft forever preferred_lft forever
```

```
#VAGRANT-BEGIN
# The contents below are automatically generated by Vagrant. Do not modify.
auto eth0
iface eth0 inet static
      address 192.168.56.10
      netmask 255.255.255.0
#VAGRANT-END
```

## vagrant reload 2

site unavailable; pings fail. `/etc/network/interfaces` looks right, but eth0 still has the ip and the nic is now completely disabled in vbox (shows as disabled in vbox gui)

```
$ vagrant ssh -- ip a

1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 08:00:27:d1:30:c3 brd ff:ff:ff:ff:ff:ff
    inet 10.0.2.15/24 brd 10.0.2.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet 192.168.56.10/24 brd 192.168.56.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::a00:27ff:fed1:30c3/64 scope link
       valid_lft forever preferred_lft forever
3: eth1: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN group default qlen 1000
    link/ether 08:00:27:82:2b:6b brd ff:ff:ff:ff:ff:ff
```

```
$ vagrant ssh -- cat /etc/network/interfaces

#VAGRANT-BEGIN
# The contents below are automatically generated by Vagrant. Do not modify.
auto eth0
iface eth0 inet static
      address 192.168.56.10
      netmask 255.255.255.0
#VAGRANT-END
```

## vagrant reload 3

can't ping guest

```
$ vagrant ssh -- ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 08:00:27:d1:30:c3 brd ff:ff:ff:ff:ff:ff
    inet 10.0.2.15/24 brd 10.0.2.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet 192.168.56.10/24 brd 192.168.56.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::a00:27ff:fed1:30c3/64 scope link
       valid_lft forever preferred_lft forever
3: eth1: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN group default qlen 1000
    link/ether 08:00:27:82:2b:6b brd ff:ff:ff:ff:ff:ff
```

```
$ vagrant ssh -- cat /etc/network/interfaces

...
#VAGRANT-BEGIN
# The contents below are automatically generated by Vagrant. Do not modify.
auto eth0
iface eth0 inet static
      address 192.168.56.10
      netmask 255.255.255.0
#VAGRANT-END
```

**Comment by jamiejackson** [comment]
Will work on a repro case.

**Comment by jamiejackson** [comment]
...in the meantime, I've also found that it's a problem in 1.9, too. I'm trying 1.8 now, since a colleague who hasn't upgraded in a long time (and has 1.8.1) has never seen the problem.

**Comment by ZoranStojsavljevic** [comment]
Hello Jamie,

Thank you for joining the club.

Since I had these problems (similar as yours), and lost at least three full working days to solve this problem (since from the very beginning I installed Vagrant .rpm, appeared to be version 2.0.1, at the end, not to degrade the Vagrant version, I did the following:
[1] introduced the only one option (except port forwarding) for the networking, which is shown as last line in the networking script:
```
  ## Vagrant network configuration ==>> MUST stay as original!?

  # Forward port 8888 for the internal REST server
  config.vm.network :forwarded_port, guest: 8888, host: 8888
  ## Forward port 8020 for the ser2net terminal
  ## config.vm.network :forwarded_port, guest: 8020, host: 8020
  # Forward port 8010 for the Storage Server
  config.vm.network :forwarded_port, guest: 8010, host: 8010
  # Forward port 5000 for the KernelCI Frontend Web Server
  config.vm.network :forwarded_port, guest: 5000, host: 5000
  # Forward port 80 for the http Lava Frontend Web Server
  config.vm.network :forwarded_port, guest: 8080, host: 8080
  # Forward port 443 for the https Lava Frontend Web Server
  config.vm.network :forwarded_port, guest: 443, host: 4443
  # Configure network accessibility for tftp server
  config.vm.network "public_network", use_dhcp_assigned_default_route: true
```
Which gives me in the VirtualBox VM the following option:
```
root@stretch:~# ip route
10.0.2.0/24 dev eth0 proto kernel scope link src 10.0.2.15 
root@stretch:~#
```
And then, I run /bin/bash shell script to create all the networking I need (parts shown):
```
## [1] Check if there is a route to www/0.0.0.0, so the necessary setup could be performed
T_VAR=`ip route show | grep default | awk '{print $1}'`

echo "default route installed? $T_VAR"

if [ "$T_VAR" != "default" ]; then
	## More code to check eth0 i/f and specific VBox gw IP address must be added here
	ip route add default via 10.0.2.2 dev eth0
	sleep 1 ## wait for ip route add default to perform!
	T_VAR=`ip route show | grep default | awk '{print $1}'`
	if [ "$T_VAR" != "default" ]; then
		exit 127
	fi
fi

## Test along the execution
ip route show > /dev/null

## [2] Test eth2 interface, since this one (so far) is used as pass-through!

T_VAR=`ifconfig -a | grep "eth2" | awk '{print $1}' | sed 's/://g'`

if [ "$T_VAR" != "eth2" ]; then
        echo "VMM USB/ETH passthrough mechanism failed!"
        exit 127
fi

T_VAR=`ifconfig -a eth2 | grep 'inet ' | awk '{print $2}'`

if [ -z "$T_VAR" ]; then
        ## give to eth2 the formal static address for DHCP init
        ifconfig eth2 192.168.15.2 up
fi

## Test along the execution
ifconfig eth2 > /dev/null
```
I guess, with Vagrant CORRECT networking behaviour, namely ONLY two lines:
```
config.vm.network "public_network", use_dhcp_assigned_default_route: true
config.vm.network "private_network", bridge: "eth2", ip: "192.168.15.2", type: "static"
```
I should achieve the following in Virtual Box VM WITHOUT having above /bin/bash part of the script:
```
root@stretch:~# ip route
default via 10.0.2.2 dev eth0 
10.0.2.0/24 dev eth0 proto kernel scope link src 10.0.2.15 
192.168.15.0/24 dev eth2 proto kernel scope link src 192.168.15.2 
root@stretch:~# 
```
And then reduce my /bin/bash scripting to in.tftpd, dnsmasq and other services I need.

Thank you,
Zoran

**Comment by jamiejackson** [comment]
Thanks @ZoranStojsavljevic, I'll take a closer look at your reply.

BTW, in trying to distill my problem into a repro case, I found that some combination of things in my docker stack (I'm using the docker provider) is contributing to it. It takes a long time to elicit the problem (fresh vagrant up plus two reloads) so it takes me a long time to iterate after making a change to see if the problem still exists, so it's making this difficult.

**Comment by jamiejackson** [comment]
Repro case: https://github.com/jamiejackson/vagrant-bug-9546

**Comment by kikitux** [comment]
Well, maybe its a bug in vagrant-docker-compose plugin ?

**Comment by ZoranStojsavljevic** [comment]
> Well, maybe its a bug in vagrant-docker-compose plugin ?

Not really. I do NOT run any docker. And I have plenty of problems with Vagrant networking, as described per above (I run my own networking script as provisioner, which substitutes Vagrant networking part).

Zoran

**Comment by jamiejackson** [comment]
The docker compose provisioner ultimately just invokes docker commands. I
think it more likely that my example is eliciting an edge case bug of
vagrant, itself.

It seems as though the addition of other (unrelated) virtual NICs can cause
vagrant to lose track of which interface is which--in my case, detecting
that eth0 (normally NAT) is *also* the private network (normally eth1).

**Comment by jamiejackson** [comment]
I'm not completely satisfied with this workaround, but it seems to prevent the problem and gets my team working without disruption for now. (I still suspect I'm working around a bug with vagrant getting confused about which NIC is which.)

Side note: Due to a bug (I'd call it a bug, anyway - https://github.com/emyl/vagrant-triggers/issues/87 ) in vagrant-triggers, I've instructed the team to use `vagrant halt && vagrant up` instead of `vagrant reload`, since I can't seem to attach triggers to reload actions.

```ruby
  config.trigger.before :halt do
    info "Remove docker containers and networks to prevent future network probs" 
    run_remote  "docker stop $(docker ps -aq) || true \
      && docker rm $(docker ps -aq) || true \
      && docker network prune --force"
    info "In case of an accidental `vagrant reload`, clean up NAT interface."
    run_remote "ip addr del #{ip}/24 dev eth0 || true"
  end
```
In other words, before halting, remove all docker networks, so next time I `vagrant up`, vagrant won't get confused about which NIC is the vagrant-managed, private one. [The following `vagrant up` will re-generate the containers (from cache) and networks quickly.]

**Comment by jamiejackson** [comment]
By the way, I think one of the more compelling clues that this is a vagrant bug is this (from my OP).

From `/etc/network/interfaces`, after a reload:
```
#VAGRANT-BEGIN
# The contents below are automatically generated by Vagrant. Do not modify.
auto eth0
iface eth0 inet static
      address 192.168.56.10
      netmask 255.255.255.0
#VAGRANT-END
```
That's vagrant using `eth0` as the "private" NIC (caught red-handed). It seems far-fetched to implicate anything else as the _root_ cause.

**Comment by causton81** [comment]
fwiw, the Vagrantfile below always configures eth2 with the same IP as eth1. **Moving forwarded_port after the private_network lines works as expected** (or just removing forwarded_port altogether).
```
Vagrant.configure("2") do |config|
  config.vm.box = "centos/7"
  config.vm.box_version = "1804.02"
  config.vm.define "vm1" do |sub|
    sub.vm.network "forwarded_port", guest: 80, host: 8080
    sub.vm.network "private_network", ip: "192.168.33.10"
    sub.vm.network "private_network", ip: "192.168.34.10"
  end
end
```

```
~/repro$ vagrant version
Installed Version: 2.1.5
Latest Version: 2.1.5

You're running an up-to-date version of Vagrant!
```
```
~/repro$ vagrant ssh -c 'ip ad|grep 192.168.33.10'
    inet 192.168.33.10/24 brd 192.168.33.255 scope global noprefixroute eth1
    inet 192.168.33.10/24 brd 192.168.33.255 scope global noprefixroute eth2
Connection to 127.0.0.1 closed.
```

**Comment by lurdan** [comment]
hi, I've met similar problem (setting private_network makes 2nd vagrant-up fail).

In my case, vagrant fails to execute ifdown at:
https://github.com/hashicorp/vagrant/blob/4d4e707b73d3d4aecd80bbe8f38cce5eb0e4268b/plugins/guests/debian/cap/configure_networks.rb#L142

(I guess eth1 is configured without using ifup, but I'm not sure...)

My dirty workaround, adding '--force' to ifdown command seems to solve the problem.
How about your case?

windows10 / vagrant 2.2.2 / virtualbox 5.2.20 / debian stretch as guest vm

**Comment by lurdan** [comment]
@chrisroberts Thanks, it turns out that with basic Vagrantfile with docker provision, vagrant runs fine repeatedly.
So, our local provisioning seems to corrupt them. My case is not relevant to this issue.
